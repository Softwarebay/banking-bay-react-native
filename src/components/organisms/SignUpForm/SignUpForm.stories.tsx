import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import { ViewStory } from 'utils/storybook/ViewStory';

import { SignUpForm } from './SignUpForm';

storiesOf('Organisms/SignUpForm', module)
	.addDecorator((getStory) => <ViewStory>{getStory()}</ViewStory>)
	.add('Default', () => (
		<SignUpForm onSubmitSignUpForm={action('onSubmitSignUpForm')} />
	));
