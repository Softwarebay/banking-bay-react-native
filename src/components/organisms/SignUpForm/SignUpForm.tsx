import { useEffect, useRef, useState } from 'react';
import { View } from 'react-native';

import { Button, InputFieldHandle } from 'components/atoms';

import { EmailInput, PasswordInput, TextInput } from 'components/molecules';
import { styles } from './SignUpForm.styles';

export interface Props {
	onSubmitSignUpForm: (userCredentials: {
		name: string;
		email: string;
		password: string;
	}) => void;
}

export const SignUpForm = ({ onSubmitSignUpForm }: Props) => {
	const nameRef = useRef<InputFieldHandle>(null);
	const emailRef = useRef<InputFieldHandle>(null);
	const passwordRef = useRef<InputFieldHandle>(null);
	const confirmPasswordRef = useRef<InputFieldHandle>(null);

	const [isEmailValid, setIsEmailValid] = useState(false);

	useEffect(() => {
		nameRef.current?.setValue('John Smith');
		emailRef.current?.setValue('test@test.com');
		passwordRef.current?.setValue('password');
		confirmPasswordRef.current?.setValue('password');
		setIsEmailValid(true);
	}, []);

	const validateEmail = (email: string) => {
		return !!email.match(
			/^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/
		);
	};

	const onPressSubmitButton = () => {
		const name = nameRef.current?.getValue();
		const email = emailRef.current?.getValue();
		const password = passwordRef.current?.getValue();
		const confirmPassword = confirmPasswordRef.current?.getValue();

		if (
			name &&
			email &&
			password &&
			isEmailValid &&
			password === confirmPassword
		) {
			onSubmitSignUpForm({
				name,
				email,
				password,
			});
		}
	};

	return (
		<View>
			<TextInput
				ref={nameRef}
				placeholder="Enter your name"
				style={styles.input}
				testID="sign-up-form-name-input-test-id"
			/>
			<EmailInput
				ref={emailRef}
				isValid={isEmailValid}
				placeholder="Enter your email"
				setIsValid={setIsEmailValid}
				style={styles.input}
				testID="sign-up-form-email-input-test-id"
				validate={validateEmail}
			/>
			<PasswordInput
				ref={passwordRef}
				placeholder="Enter your password"
				style={styles.input}
				testID="sign-up-form-password-input-test-id"
			/>
			<PasswordInput
				ref={confirmPasswordRef}
				placeholder="Confirm your password"
				style={styles.input}
				testID="sign-up-form-confirm-password-input-test-id"
			/>
			<Button
				testID="sign-up-form-submit-button-test-id"
				title="Sign Up"
				onPress={onPressSubmitButton}
			/>
		</View>
	);
};
