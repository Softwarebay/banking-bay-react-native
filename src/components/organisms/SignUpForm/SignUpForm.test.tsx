import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { SignUpForm } from './SignUpForm';

import type { Props } from './SignUpForm';

describe('SignUpForm', () => {
	const onSubmitSignUpForm = jest.fn<
		Props['onSubmitSignUpForm'],
		[{ name: string; email: string; password: string }]
	>();

	const props: Props = {
		onSubmitSignUpForm,
	};

	it('should match snapshot', () => {
		const tree = testRenderer.create(<SignUpForm {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render name, email, password, confirm password input fields and submit button', () => {
		render(<SignUpForm {...props} />);

		const nameInput = screen.getByTestId('sign-up-form-name-input-test-id');
		const emailInput = screen.getByTestId('sign-up-form-email-input-test-id');
		const passwordInput = screen.getByTestId(
			'sign-up-form-password-input-test-id'
		);
		const confirmPasswordInput = screen.getByTestId(
			'sign-up-form-confirm-password-input-test-id'
		);
		const submitBtn = screen.getByTestId('sign-up-form-submit-button-test-id');

		expect(nameInput).not.toBeNull();
		expect(emailInput).not.toBeNull();
		expect(passwordInput).not.toBeNull();
		expect(confirmPasswordInput).not.toBeNull();
		expect(submitBtn).toHaveTextContent('Sign Up');
	});

	it('should render with initially filled input fields', () => {
		render(<SignUpForm {...props} />);

		const nameInputField = screen.getByTestId(
			'input-field-sign-up-form-name-input-test-id'
		);
		const emailInputField = screen.getByTestId(
			'input-field-sign-up-form-email-input-test-id'
		);
		const passwordInputField = screen.getByTestId(
			'input-field-sign-up-form-password-input-test-id'
		);
		const confirmPasswordInputField = screen.getByTestId(
			'input-field-sign-up-form-confirm-password-input-test-id'
		);

		expect(nameInputField.props).toHaveProperty('value', 'John Smith');
		expect(emailInputField.props).toHaveProperty('value', 'test@test.com');
		expect(passwordInputField.props).toHaveProperty('value', 'password');
		expect(confirmPasswordInputField.props).toHaveProperty('value', 'password');
	});

	it('should submit form when input fields are filled and email is valid and passwords match', () => {
		render(<SignUpForm {...props} />);

		const submitBtn = screen.getByTestId('sign-up-form-submit-button-test-id');

		fireEvent.press(submitBtn);

		expect(onSubmitSignUpForm).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'John Smith',
				email: 'test@test.com',
				password: 'password',
			})
		);
		expect(onSubmitSignUpForm).toHaveBeenCalledTimes(1);
	});

	it("shouldn't submit form with empty inputs", () => {
		render(<SignUpForm {...props} />);

		const nameInputField = screen.getByTestId(
			'input-field-sign-up-form-name-input-test-id'
		);
		const emailInputField = screen.getByTestId(
			'input-field-sign-up-form-email-input-test-id'
		);
		const passwordInputField = screen.getByTestId(
			'input-field-sign-up-form-password-input-test-id'
		);
		const confirmPasswordInputField = screen.getByTestId(
			'input-field-sign-up-form-confirm-password-input-test-id'
		);
		const submitBtn = screen.getByTestId('sign-up-form-submit-button-test-id');

		fireEvent.changeText(nameInputField, '');
		fireEvent.changeText(emailInputField, '');
		fireEvent.changeText(passwordInputField, '');
		fireEvent.changeText(confirmPasswordInputField, '');
		fireEvent.press(submitBtn);

		expect(onSubmitSignUpForm).not.toHaveBeenCalled();
	});

	it("shouldn't submit form when input fields are filled and email is not valid", () => {
		render(<SignUpForm {...props} />);

		const emailInputField = screen.getByTestId(
			'input-field-sign-up-form-email-input-test-id'
		);
		const submitBtn = screen.getByTestId('sign-up-form-submit-button-test-id');

		fireEvent.changeText(emailInputField, 'invalidmail');
		fireEvent.press(submitBtn);

		expect(onSubmitSignUpForm).not.toHaveBeenCalled();
	});

	it("shouldn't submit form when input fields are filled and passwords don't match", () => {
		render(<SignUpForm {...props} />);

		const passwordInputField = screen.getByTestId(
			'input-field-sign-up-form-password-input-test-id'
		);
		const confirmPasswordInputField = screen.getByTestId(
			'input-field-sign-up-form-confirm-password-input-test-id'
		);
		const submitBtn = screen.getByTestId('sign-up-form-submit-button-test-id');

		fireEvent.changeText(passwordInputField, 'password');
		fireEvent.changeText(confirmPasswordInputField, 'drowssap');
		fireEvent.press(submitBtn);

		expect(onSubmitSignUpForm).not.toHaveBeenCalled();
	});
});
