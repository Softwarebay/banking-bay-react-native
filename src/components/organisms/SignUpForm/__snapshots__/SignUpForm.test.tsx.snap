// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm should match snapshot 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "height": 55,
              "paddingHorizontal": 20,
              "paddingVertical": 14,
            },
            Object {
              "marginBottom": 14,
            },
          ],
        ],
      ]
    }
    testID="sign-up-form-name-input-test-id"
  >
    <TextInput
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="Enter your name"
      secureTextEntry={false}
      style={
        Array [
          Object {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 15,
          },
          Object {
            "flex": 1,
          },
        ]
      }
      testID="input-field-sign-up-form-name-input-test-id"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "height": 55,
              "paddingHorizontal": 20,
              "paddingVertical": 14,
            },
            Object {
              "marginBottom": 14,
            },
          ],
        ],
      ]
    }
    testID="sign-up-form-email-input-test-id"
  >
    <TextInput
      autoCorrect={false}
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Enter your email"
      secureTextEntry={false}
      style={
        Array [
          Object {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 15,
          },
          Object {
            "flex": 1,
            "marginRight": 20,
          },
        ]
      }
      testID="input-field-sign-up-form-email-input-test-id"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "height": 55,
              "paddingHorizontal": 20,
              "paddingVertical": 14,
            },
            Object {
              "marginBottom": 14,
            },
          ],
        ],
      ]
    }
    testID="sign-up-form-password-input-test-id"
  >
    <TextInput
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="Enter your password"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 15,
          },
          Object {
            "flex": 1,
            "marginRight": 20,
          },
        ]
      }
      testID="input-field-sign-up-form-password-input-test-id"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "right": 0,
          },
          false,
        ]
      }
      testID="icon-btn-sign-up-form-password-input-test-id"
    >
      <SvgMock
        color="#1B1D4D"
        height={16}
        testID="eye-off-icon-test-id"
        width={16}
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "height": 55,
              "paddingHorizontal": 20,
              "paddingVertical": 14,
            },
            Object {
              "marginBottom": 14,
            },
          ],
        ],
      ]
    }
    testID="sign-up-form-confirm-password-input-test-id"
  >
    <TextInput
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="Confirm your password"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 15,
          },
          Object {
            "flex": 1,
            "marginRight": 20,
          },
        ]
      }
      testID="input-field-sign-up-form-confirm-password-input-test-id"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "right": 0,
          },
          false,
        ]
      }
      testID="icon-btn-sign-up-form-confirm-password-input-test-id"
    >
      <SvgMock
        color="#1B1D4D"
        height={16}
        testID="eye-off-icon-test-id"
        width={16}
      />
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        undefined,
      ]
    }
    testID="sign-up-form-submit-button-test-id"
  >
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4288142079,
            4279654301,
          ],
          "endPoint": Array [
            1,
            1,
          ],
          "locations": undefined,
          "startPoint": Array [
            0,
            0,
          ],
        }
      }
      style={
        Object {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 55,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "MulishSemiBold",
            "fontSize": 17,
          }
        }
      >
        Sign Up
      </Text>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
</View>
`;
